### ISSUE TITLE
BoundsError in eigs during test/arpack.jl on 32 bit

### ISSUE THREAD

**Issue by tkelman** [JuliaLang/LinearAlgebra.jl#132]
This happens on both 32 bit Linux (compiled within 64 bit Ubuntu by setting `ARCH=i686` in Make.user) and 32 bit Windows, reasonably often. The backtrace is more useful on Linux:

```
tkelman@ygdesk:~/Julia/julia-linux32$ ./julia -e 'versioninfo()'
Julia Version 0.3.0-rc1+186
Commit 542b013 (2014-07-25 04:08 UTC)
Platform Info:
  System: Linux (i686-linux-gnu)
  CPU: Intel(R) Core(TM)2 Duo CPU     E8400  @ 3.00GHz
  WORD_SIZE: 32
  BLAS: libopenblas (DYNAMIC_ARCH NO_AFFINITY Penryn)
  LAPACK: libopenblas
  LIBM: libopenlibm
  LLVM: libLLVM-3.3
tkelman@ygdesk:~/Julia/julia-linux32$ make test-arpack
    JULIA test/arpack
     * arpack
exception on 1: ERROR: BoundsError()
 in checkbounds at abstractarray.jl:62
 in setindex! at multidimensional.jl:61
 in eupd_wrapper at linalg/arpack.jl:174
 in eigs at linalg/arnoldi.jl:105
 in eigs at linalg/arnoldi.jl:5
 in runtests at /home/tkelman/Julia/julia-linux32/test/testdefs.jl:5
 in anonymous at multi.jl:660
 in run_work_thunk at multi.jl:621
 in remotecall_fetch at multi.jl:694
 in remotecall_fetch at multi.jl:709
 in anonymous at task.jl:1363
while loading arpack.jl, in expression starting on line 102
ERROR: BoundsError()
 in checkbounds at abstractarray.jl:62
 in setindex! at multidimensional.jl:61
 in eupd_wrapper at linalg/arpack.jl:174
 in eigs at linalg/arnoldi.jl:105
 in eigs at linalg/arnoldi.jl:5
 in runtests at /home/tkelman/Julia/julia-linux32/test/testdefs.jl:5
 in anonymous at multi.jl:660
 in run_work_thunk at multi.jl:621
 in remotecall_fetch at multi.jl:694
 in remotecall_fetch at multi.jl:709
 in anonymous at task.jl:1363
while loading arpack.jl, in expression starting on line 102
while loading /home/tkelman/Julia/julia-linux32/test/runtests.jl, in expression starting on line 46

make[1]: *** [arpack] Error 1
make: *** [test-arpack] Error 2
```

cc @ViralBShah @jutho

**Comment by Jutho** [comment]
There is indeed nothing explicitly preventing line 174 to execute for `j=nev+1`, although this should not happen. Will the ARCH=i686 also work on OS X? I will try to take a look at this.

**Comment by tkelman** [comment]
I'm not sure whether all the cross-compilation for 32 bit will work on Mac, but it might? Best to try in a new fresh checkout. If it doesn't work on a Mac and you don't have 32 bit Linux handy on a real machine or a VM, I think @staticfloat can get you access to one of his VM's.

**Comment by staticfloat** [comment]
Hmmmm.  On my 32-bit Ubuntu VM, I don't see this:

```
$ git log -1
commit 542b013eb6dd08480ee11da1db2449d13bfa39e2
Merge: 2f26857 1f09d55
Author: Jeff Bezanson <jeff.bezanson@gmail.com>
Date:   Fri Jul 25 00:08:47 2014 -0400

    Merge pull request JuliaLang/julia#7713 from amitmurthy/amitm/7646

    addprocs: wait till all workers are connected to each other.
$ ../julia runtests.jl arpack
     * arpack
    SUCCESS
```

I wonder if this is an `ARCH=i686`-specific issue?

**Comment by staticfloat** [comment]
Note that I'm using distribution packages for almost all dependencies there.  I'll try just a plain `make` and see if I get the same thing.

**Comment by ViralBShah** [comment]
I do have a 32-bit ubuntu vm and will get to this next week.

**Comment by staticfloat** [comment]
Nope.  `make` from scratch doesn't see this error.

**Comment by tkelman** [comment]
It's intermittent, so be sure to run several times. It does happen with Win32 Julia, which I guess also shares some aspects of being 32-on-64.

Edit: you might need `--check-bounds=yes`, I'm not sure if any of the eigs code uses `@inbounds`. And this definitely seems more difficult to trigger on an actual 32 bit machine than 32-on-64, either Windows or Linux. Yet another example of Julia getting caught out by the intricacies of x87 floating point behavior?

**Comment by staticfloat** [comment]
Aha!  I got it;

```
$ while true; do ../julia --check-bounds=yes runtests.jl arpack; done
     * arpack
    SUCCESS
     * arpack
    SUCCESS
     * arpack
exception on 1: ERROR: BoundsError()
 in checkbounds at abstractarray.jl:62
 in setindex! at multidimensional.jl:61
 in eupd_wrapper at linalg/arpack.jl:174
 in eigs at linalg/arnoldi.jl:105
 in eigs at linalg/arnoldi.jl:5
 in runtests at /home/vagrant/julia/test/testdefs.jl:5
 in anonymous at multi.jl:660
 in run_work_thunk at multi.jl:621
 in remotecall_fetch at multi.jl:694
 in remotecall_fetch at multi.jl:709
 in anonymous at task.jl:1363
while loading arpack.jl, in expression starting on line 102
ERROR: BoundsError()
 in checkbounds at abstractarray.jl:62
 in setindex! at multidimensional.jl:61
 in eupd_wrapper at linalg/arpack.jl:174
 in eigs at linalg/arnoldi.jl:105
 in eigs at linalg/arnoldi.jl:5
 in runtests at /home/vagrant/julia/test/testdefs.jl:5
 in anonymous at multi.jl:660
 in run_work_thunk at multi.jl:621
 in remotecall_fetch at multi.jl:694
 in remotecall_fetch at multi.jl:709
 in anonymous at task.jl:1363
while loading arpack.jl, in expression starting on line 102
while loading /home/vagrant/julia/test/runtests.jl, in expression starting on line 46
```

I'll just leave the VM in the state it's in right now in case Viral wants to take a peek.

**Comment by ViralBShah** [comment]
Where is this VM? Could you email me instructions on accessing it?

**Comment by ViralBShah** [comment]
The bounds error can happen potentially at line 175 in arpack.jl. Figuring out how best to avoid it.

**Comment by ViralBShah** [comment]
cc: @Jutho

**Comment by ViralBShah** [comment]
Fixed by JuliaLang/julia#7741
